IS1001Test.py

```diff
@@ -12,12 +12,623 @@
+
+        status, response = self._make_auth_request(
+            method="POST", url_path='register_client', data=request_data, auth="user"
+        )
+
+        self._verify_response(test=test, expected_status=201, response=response)
+
+        for key in request_data:
+            # Test that all request data keys are found in response. Added 's' compensates for grant_type/s, etc.
+            if not any(i in [key, key + 's'] for i in response.json().keys()):
+                return test.FAIL(
+                    """'{}' value not in response keys.
+                    'The authorization server MUST return all registered metadata about the client'""".format(key)
+                )
+            # Check that same keys in req and resp have same values
+            if key in response.json().keys() and request_data[key] != response.json()[key]:
+                return test.WARNING("The response for {} did not match. Request: {} != Response: {}".format(
+                    key, request_data[key], response.json()[key])
+                )
+
+        # Test that all array fields in response contain a value from the corresponding value in the request
+        for field in ARRAY_RESPONSE_FIELDS:
+            if field in request_data:
+                index = field
+            elif field[0:-1] in request_data:
+                index = field[0:-1]
+            else:
+                continue
+            # This is in case url-encoded form data is used instead of JSON
+            if isinstance(request_data[index], str):
+                request_data[index] = request_data[index].split(LIST_DELIMITER)
+            if not all(i in request_data[index] for i in response.json()[field]):
+                return test.FAIL("'{}' value is not included in the response from the Auth Server. {} not in {}"
+                                 .format(field, request_data[index], response.json()[field]))
+
+        if response.json()["token_endpoint_auth_method"] in ["none", None, "None", "null"]:
+            return test.WARNING("Token Endpoint Authorization method SHOULD NOT be None")
+
+        if not isinstance(response.json()["scope"], str):
+            return test.FAIL("Scope values MUST be a space-delimited string")
+
+        if not set(RECOMMENDED_RESPONSE_FIELDS).issubset(response.json().keys()):
+            return test.WARNING("Client registration response SHOULD include: {}".format(RECOMMENDED_RESPONSE_FIELDS))
+
+        register_client_schema = self.get_schema(AUTH_API_KEY, "POST", '/register_client', 201)
+
+        try:
+            self.validate_schema(response.json(), register_client_schema)
+            if response.status_code == 201:
+                self.client_data = response.json()
+                return test.PASS()
+        except Exception as e:
+            return test.FAIL("Status code was {} and Schema validation failed. {}".format(response.status_code, e))
+
+        return test.FAIL("Implementation did not pass the necessary criteria.")

```

Comment:
Could this error message be improved at all? As far as I can see you only get here if the final response status code wasn't a 201, so perhaps a 200, which this wouldn't help to debug.

Reply:
This was meant to be a "I've tested for everything I can think of and its still not passing so who knows"... the `validate_schema` call should flag an error if the status code isn't 201 (kinda rendering the conditional `if` statement pointless but its god to be explicit)

Reply:
Ah, did you mean '_verify_response'? I'd suggest removing the 201 check here given it is a duplicate.